src/main/java/io/supertokens/webserver/api/passwordless/ConsumeCodeAPI.java

```diff
@@ -0,0 +1,121 @@
+
+import java.io.IOException;
+import java.security.InvalidKeyException;
+import java.security.NoSuchAlgorithmException;
+
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import com.google.gson.Gson;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParser;
+
+import io.supertokens.Main;
+import io.supertokens.passwordless.Passwordless;
+import io.supertokens.passwordless.Passwordless.ConsumeCodeResponse;
+import io.supertokens.passwordless.exceptions.Base64EncodingException;
+import io.supertokens.passwordless.exceptions.DeviceIdHashMismatchException;
+import io.supertokens.passwordless.exceptions.ExpiredUserInputCodeException;
+import io.supertokens.passwordless.exceptions.IncorrectUserInputCodeException;
+import io.supertokens.passwordless.exceptions.RestartFlowException;
+import io.supertokens.pluginInterface.RECIPE_ID;
+import io.supertokens.pluginInterface.exceptions.StorageQueryException;
+import io.supertokens.pluginInterface.exceptions.StorageTransactionLogicException;
+import io.supertokens.webserver.InputParser;
+import io.supertokens.webserver.WebserverAPI;
+
+public class ConsumeCodeAPI extends WebserverAPI {
+
+    private static final long serialVersionUID = -4641988458637882374L;
+
+    public ConsumeCodeAPI(Main main) {
+        super(main, RECIPE_ID.PASSWORDLESS.toString());
+    }
+
+    @Override
+    public String getPath() {
+        return "/recipe/signinup/code/consume";
+    }
+
+    @Override
+    protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws IOException, ServletException {
+        // Logic based on: https://app.code2flow.com/OFxcbh1FNLXd
+        JsonObject input = InputParser.parseJsonObjectOrThrowError(req);
+
+        String linkCode = null;
+        String deviceId = null;
+        String userInputCode = null;
+
+        String deviceIdHash = InputParser.parseStringOrThrowError(input, "preAuthSessionId", false);
+
+        if (input.has("linkCode")) {
+            if (input.has("userInputCode") || input.has("deviceId")) {
+                throw new ServletException(
+                        new BadRequestException("Please provide exactly one of linkCode or deviceId+userInputCode"));
+            }
+            linkCode = InputParser.parseStringOrThrowError(input, "linkCode", false);
+        } else if (input.has("userInputCode") && input.has("deviceId")) {
+            deviceId = InputParser.parseStringOrThrowError(input, "deviceId", false);
+            userInputCode = InputParser.parseStringOrThrowError(input, "userInputCode", false);
+        } else {
+            throw new ServletException(
+                    new BadRequestException("Please provide exactly one of linkCode or deviceId+userInputCode"));
+        }
+
+        try {
+            ConsumeCodeResponse consumeCodeResponse = Passwordless.consumeCode(main, deviceId, deviceIdHash,
```

Comment:
In this function, if I pass the `linkCode`, even without the `=` at the end of it, for example instead of `FWFTVYIj7z8MVtWZn502cklRGsy5oWr9ydp_VGYmJW8=`, I pass `FWFTVYIj7z8MVtWZn502cklRGsy5oWr9ydp_VGYmJW8`, it still works. Is that expected?

Reply:
I think we talked about it previously, but it didn't make it into a comment, so: It is expected: the = is just padding, and it looks like the java encoder doesn't strip the "=" while URL encoding but treats them as optional when parsing.